### PR TITLE
fixtures: remove s3_fake_creds_file, allow for finer boto3 client configuration

### PR DESCRIPTION
**Pull request by dtrifiro**
By passing `client_kwargs` to the `s3` temp upath factory, we can avoid having to monkeypatch s3 environment variable for the whole test session.

### CODE REVIEW

File: tests/test_fixtures.py
Comment (shcheklein) [inline tests/test_fixtures.py:19]: Q: how about profile, and security token? do we need to tests for them?

Comment (shcheklein) [reply tests/test_fixtures.py:19]: do we need to include the into the "client_kwargs" dict ?

File: src/pytest_servers/s3.py
Comment (shcheklein) [inline src/pytest_servers/s3.py:53]: C: in can be used also to set env var per test (e.g. creating a function scoped fixture). E.g. in my case I don't have access directly to S3 boto. So, I would still need to put them into evn or credentials file

Comment (shcheklein) [reply src/pytest_servers/s3.py:53]: ... or pass them into DVC config - that might work as well in my case

Comment (shcheklein) [reply src/pytest_servers/s3.py:53]: I guess, actionable discussion here / thought is if we want to have an extra fixture, function scoped that sets env vars.... probably it's fine to leave it to the client

**Review by shcheklein — COMMENTED** [review]
Thanks @dtrifiro ! I'm fine with the change (I asked a few questions to clarify). Will it break things on the dvc-s3 side / dvc - they might expect those secrets to be set already ... I guess we can check that by waiting for renovate to pick up an update

File: tests/test_fixtures.py
Comment (dtrifiro) [reply tests/test_fixtures.py:19]: The main reason why I ended up going for this `client_kwargs` dict is that by looking at the boto3 [credentials docs](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html), it mentions that the order it uses to search for credentials is the following:

1. **Passing credentials as parameters in the boto.client() method**
2. Passing credentials as parameters when creating a Session object
3. Environment variables
4. Shared credential file (~/.aws/credentials)
5. AWS config file (~/.aws/config)
6. Assume Role provider
7. Boto2 config file (/etc/boto.cfg and ~/.boto)
8. Instance metadata service on an Amazon EC2 instance that has an IAM role configured.

This means that if we pass `aws_session_token` `aws_secret_access_key` and `aws_access_key_id`, it should be enough:

- security token is an environment variable and has lower priority. From the docs 
```The AWS_SECURITY_TOKEN environment variable can also be used, but is only supported for backwards compatibility purposes. ```
- Profile is used to parse session token, secret access key and id from the .aws credentials/config files, which also have lower priority


Regarding the `client_kwargs` dict, we could make it more explicit and pass `key`, `secret`, `token` etc as kwargs, since UPath passes these directly as `S3FileSystem` init kwargs (https://s3fs.readthedocs.io/en/latest/api.html#s3fs.core.S3FileSystem), but I was a bit concerned in introducing yet another adapter, plus with client_kwargs we could eventually configure the underlying client with options that S3FileSystem hasn't exposed yet

File: src/pytest_servers/s3.py
Comment (dtrifiro) [reply src/pytest_servers/s3.py:53]: > E.g. in my case I don't have access directly to S3 boto. So, I would still need to put them into evn or credentials file

Depending on what you're trying to accomplish, you can use the `fs` attribute of the upath instance to access `S3FileSystem` and eventually even use the underlying client (although it's async, so your mileage may vary)

```python

def test_something(tmp_s3_path):
    fsspec_fs = tmp_s3_path.fs
    aiobotocore_client = fsspec_fs
    ...
```

Regarding the fixture to set env vars: maybe we can work something out in another PR, I have some code but it has issues with fixture scopes.

**Comment by dtrifiro** [comment]
> Will it break things on the dvc-s3 side

Hopefully not :smile:: the API is unchanged, and the upath should have all it needs to properly communicate with the moto server

**Comment by shcheklein** [comment]
> Hopefully not 😄: the API is unchanged, and the upath should have all it needs to properly communicate with the moto server

yep, just thinking if some code was relying on those env vars present. Anyways, we can find out that.

File: tests/test_fixtures.py
Comment (shcheklein) [reply tests/test_fixtures.py:19]: Good. Make sense.

File: src/pytest_servers/s3.py
Comment (shcheklein) [reply src/pytest_servers/s3.py:53]: There are plenty of tests where we won't have access to boto, etc to pass the config directly I think.

And example of a test is you use dvc repo api to push / pull. The only think you control is dvc remote config or env vars, there is not direct access to boto usually. 

 But I also think it's fine for now. Client can put together a simple fixture to override env, or set DVC config or some other config. It's way more flexible now with these all values available.

I can try to use it after we merge and let you know if that works or not.

**Review by shcheklein — APPROVED** [review]

**Comment by shcheklein** [comment]
@efiop @dtrifiro @skshetry can we merge and release this please? :)

**Comment by efiop** [comment]
Thanks @dtrifiro ! 🙏

**Comment by jonburdo** [comment]
Looks like we did have [some code]( https://github.com/iterative/dql/pull/1086 ) that relied on the env vars and I'd expect that to be normal. It's probably reasonable for users to set them if they need them, although if they create a path with `tmp_upath_factory.mktemp`, they'll need to know to take values from `path.fs.client_kwargs`

**Pull request merged**
